app/models/concerns/has_coordinates.rb

```diff
@@ -2,33 +2,34 @@ module HasCoordinates
   extend ActiveSupport::Concern
 
   included do
-    geocoded_by :full_address
-    after_validation :geocode, if: ->(obj){ (obj.address.present? || obj.city.present?) and obj.full_address_changed? }
-    after_validation :location_found?
+    after_validation :geocode, if: :geocodable?
   end
 
   def full_address
-    [address, city].reject(&:blank?).join(' ')
+    full_address = [address, city].reject(&:blank?)
+    return if full_address.none? 
+
+    full_address.join(' ')
   end
 
   def full_address_changed?
-    :address_changed? || :city_changed?
+    address_changed? || city_changed?
   end
 
-  # Check if location was found by geocoder.
-  # If full address is blank, set coords to nil
-  # If full adress has changed, but coords did not, it means geocoder was unable to find new location
-  def location_found?
-    if (full_address.blank?)
+  def geocode
+    results = Geocoder.search(full_address)
+    if results.empty?
```

Comment:
you can probably solve this a little slimmer with safe navigation:

```ruby
self.latitude = results&.first&.latitude || nil
self.longitude = results&.first&.longitude || nil
```

Reply:
But it is a nitpick. If you find your solution more readable. Stick with what you have. I basically only wanted to mention the option.